pkg/cmdline/cmd.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package cmdline
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// CommandManager holds root command or first parent
+// and can stores group of command
+type CommandManager struct {
+	rootCmd   *cobra.Command
+	groupCmds map[string][]*cobra.Command
+	errPool   []error
+	fm        *flagManager
+}
+
+// FlagError represents an error type returnes when command
+// line parsing fails
+type FlagError string
+
+func (f FlagError) Error() string {
+	return string(f)
+}
+
+func onError(cmd *cobra.Command, err error) error {
+	return FlagError(err.Error())
+}
+
+// NewCommandManager instantiates a CommandManager
+func NewCommandManager(rootCmd *cobra.Command) *CommandManager {
+	if rootCmd == nil {
+		panic("nil root command passed")
+	}
+	cm := &CommandManager{
+		rootCmd:   rootCmd,
+		groupCmds: make(map[string][]*cobra.Command),
+		fm:        newFlagManager(),
+	}
+	rootCmd.SetFlagErrorFunc(onError)
+	return cm
+}
+
+func (m *CommandManager) isRegistered(cmd *cobra.Command) bool {
+	c := cmd.Parent()
+	for c != nil {
+		if c == m.rootCmd {
+			break
+		}
+		c = c.Parent()
+	}
+	return c != nil
+}
+
+func (m *CommandManager) pushError(err error) {
+	m.errPool = append(m.errPool, err)
+}
+
+// GetError returns the error pool
+func (m *CommandManager) GetError() []error {
+	return m.errPool
+}
+
+// RegisterCmd registers a child command for the root command
+func (m *CommandManager) RegisterCmd(cmd *cobra.Command, interspersed bool) {
+	// panic here because it's a misuse of API and generally from
+	// global context or init() functions
+	if cmd == nil {
+		panic("nil command passed")
+	}
+	cmd.SetFlagErrorFunc(onError)
+	m.rootCmd.AddCommand(cmd)
+	cmd.Flags().SetInterspersed(interspersed)
+	m.SetCmdGroup(m.GetCmdName(cmd), cmd)
+}
+
+// RegisterSubCmd registers a child command for parent command given as argument
+func (m *CommandManager) RegisterSubCmd(parentCmd, childCmd *cobra.Command, interspersed bool) {
+	// panic here because it's a misuse of API and generally from
+	// global context or init() functions
+	if parentCmd == nil {
+		panic("nil parent command passed")
+	} else if childCmd == nil {
+		panic("nil child command passed")
+	} else if !m.isRegistered(parentCmd) {
+		panic("parent command not registered")
+	}
+	parentCmd.AddCommand(childCmd)
+	childCmd.Flags().SetInterspersed(interspersed)
+	m.SetCmdGroup(m.GetCmdName(childCmd), childCmd)
+}
+
+// SetCmdGroup creates a group of commands identified by name
+func (m *CommandManager) SetCmdGroup(name string, cmds ...*cobra.Command) {
+	if m.groupCmds[name] != nil {
+		panic(fmt.Sprintf("group %s already exists", name))
```

Comment:
This function create a group in one shot manner and is not intended to be used twice for the same group, if shell command was already registered (a group "shell" was automatically created) and if someone want to create shell group it will panic.
Will move panic to a pushed error and add documentation

Reply:
Sorry but I revert to panic because I think this is the thing to do in CLI context and those panic can only be triggered by dev mistake